spring-cloud-azure-config/src/main/java/com/microsoft/azure/config/ConfigHttpClient.java

```diff
@@ -0,0 +1,128 @@
+import org.apache.http.HttpEntityEnclosingRequest;
+import org.apache.http.HttpRequest;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpUriRequest;
+import org.apache.http.client.utils.URIBuilder;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+
+import static org.apache.commons.codec.digest.HmacAlgorithms.HMAC_SHA_256;
+import static org.apache.commons.codec.digest.MessageDigestAlgorithms.SHA_256;
+
+/**
+ * Util class to execute http request, before sending http request, valid request headers
+ * will be added for each request based on given credential ID and secret.
+ *
+ * How to use:
+ * <p>
+ * HttpGet httpGet = new HttpGet("https://my-config-store.azconfig.io/keys");
+ * CloseableHttpResponse response = ConfigHttpClient.execute(httpGet, "my-credential",
+ * "my-secret");
+ * <p/>
+ */
+@Slf4j
+public class ConfigHttpClient {
+    private static final String DATE_FORMAT = "EEE, d MMM yyyy hh:mm:ss z";
+
+    private static final SimpleDateFormat GMT_DATE_FORMAT = new SimpleDateFormat(DATE_FORMAT);
+
+    static {
+        GMT_DATE_FORMAT.setTimeZone(TimeZone.getTimeZone("GMT"));
+    }
+
+    public static CloseableHttpResponse execute(HttpUriRequest request, String credential, String secret)
+            throws IOException, URISyntaxException {
+        Map<String, String> authHeaders = buildRequestHeaders(request, new Date(), credential, secret);
+        authHeaders.forEach(request::setHeader);
+
+        try (CloseableHttpClient httpClient = HttpClients.createDefault()) {
+            return httpClient.execute(request);
+        }
+    }
+
+    private static Map<String, String> buildRequestHeaders(HttpUriRequest request, Date date, String credential,
+            String secret) throws URISyntaxException, IOException {
+        String requestTime = GMT_DATE_FORMAT.format(date);
+
+        String content = "";
+        if (request instanceof HttpEntityEnclosingRequest) {
+            content = copyInputStream(((HttpEntityEnclosingRequest) request).getEntity().getContent());
+        }
+
+        byte[] digest = new DigestUtils(SHA_256).digest(content);
+        String contentHash = Base64.getEncoder().encodeToString(digest);
+
+        // SignedHeaders
+        String signedHeaders = "x-ms-date;host;x-ms-content-sha256";
+
+        // String-To-Sign
+        String methodName = request.getRequestLine().getMethod().toUpperCase();
+        String requestPath = getPath(request);
+        String host = getHost(request);
+        String toSign = String.format("%s\n%s\n%s;%s;%s", methodName, requestPath, requestTime, host, contentHash);
+
+        // Signature
+        byte[] decodedKey = Base64.getDecoder().decode(secret);
+        String signature = encodeHmac(HMAC_SHA_256, decodedKey, toSign);
+
+        // Compose headers
+        Map<String, String> headers = new HashMap<>();
+        headers.put("x-ms-date", requestTime);
```

Comment:
For some constant, suggest to put it into constant field.

Reply:
As those constants are not frequently reused, will not put into extra constant field.